solver/src/liquidity/uniswap.rs

```diff
@@ -85,89 +93,77 @@ impl UniswapLikeLiquidity {
                 settlement_handling: self.inner.clone(),
             })
         }
-        self.cache_allowances(tokens.into_iter()).await;
+        self.cache_allowances(tokens).await?;
         Ok(result)
     }
 
-    async fn cache_allowances(&self, tokens: impl Iterator<Item = H160>) {
-        let mut batch = CallBatch::new(self.web3.transport());
-        let results: Vec<_> = tokens
-            .map(|token| {
-                let allowance = ERC20::at(&self.web3, token)
-                    .allowance(
-                        self.inner.gpv2_settlement.address(),
-                        self.inner.router.address(),
-                    )
-                    .batch_call(&mut batch);
-                (token, allowance)
-            })
-            .collect();
-
-        let _ = batch.execute_all(MAX_BATCH_SIZE).await;
-
-        // await before acquiring lock so we can use std::sync::mutex (async::mutex wouldn't allow AmmSettlementHandling to be non-blocking)
-        let mut token_and_allowance = Vec::with_capacity(results.len());
-        for (pair, allowance) in results {
-            token_and_allowance.push((pair, allowance.await.unwrap_or_default()));
-        }
+    async fn cache_allowances(&self, tokens: HashSet<H160>) -> Result<()> {
+        let router = self.inner.router.address();
+        let allowances = self
+            .settlement_allowances
+            .get_allowances(tokens, router)
+            .await?;
 
-        self.inner
+        *self
```

Comment:
Wouldn't this lead to forgetting previously cached allowances? Shouldn't we keep the logic where we extend the existing cache?

Reply:
The existing code would fetch allowances for all tokens in all pools:

https://github.com/gnosis/gp-v2-services/blob/1cbefe1357104cb073da3f26a5a9b6071dc811cc/solver/src/liquidity/uniswap.rs#L83-L96

So, while it previously extended the allowance cache, I don't think it ever made use of it as it would always re-fetch allowances for all possible tokens that can get traded and replace their values.

Reply:
Oh, I think I see the issue now - this `Inner` may be shared across runs, so it can be that an allowance that was needed before gets lost. I don't think that it would happen, but its an awfully sneaky issue based on implicit assumptions on how the `UniswapLikeLiquidity` needs to be used. 

Will update.